File: phase2/Todo app/frontend/src/services/taskService.ts
import apiClient from '@/src/lib/apiClient';

// Define the Task interface based on the API contract
export interface Task {
  id: string;
  user_id: string;
  title: string;
  description?: string;
  completed: boolean;
  created_at: string; // ISO string format
  updated_at: string; // ISO string format
}

// Define interfaces for request bodies
interface CreateTaskRequest {
  title: string;
  description?: string;
  completed?: boolean;
}

interface UpdateTaskRequest {
  title?: string;
  description?: string;
  completed?: boolean;
}

/**
 * Fetch all tasks for a user
 * @param userId The ID of the user whose tasks to fetch
 * @param completed Optional filter for completion status
 * @returns Promise resolving to an array of tasks
 */
export const fetchUserTasks = async (userId: string, completed?: boolean): Promise<Task[]> => {
  try {
    const params: Record<string, any> = {};
    if (completed !== undefined) {
      params.completed = completed;
    }
    
    const response = await apiClient.get<Task[]>(`/api/${userId}/tasks`, { params });
    return response.data;
  } catch (error: any) {
    console.error('Error fetching user tasks:', error);
    throw error;
  }
};

/**
 * Create a new task for a user
 * @param userId The ID of the user creating the task
 * @param taskData The task data to create
 * @returns Promise resolving to the created task
 */
export const createTask = async (userId: string, taskData: CreateTaskRequest): Promise<Task> => {
  try {
    const response = await apiClient.post<Task>(`/api/${userId}/tasks`, taskData);
    return response.data;
  } catch (error: any) {
    console.error('Error creating task:', error);
    throw error;
  }
};

/**
 * Get a specific task by ID
 * @param userId The ID of the user who owns the task
 * @param taskId The ID of the task to retrieve
 * @returns Promise resolving to the task
 */
export const getTaskById = async (userId: string, taskId: string): Promise<Task> => {
  try {
    const response = await apiClient.get<Task>(`/api/${userId}/tasks/${taskId}`);
    return response.data;
  } catch (error: any) {
    console.error('Error fetching task:', error);
    throw error;
  }
};

/**
 * Update a task
 * @param userId The ID of the user who owns the task
 * @param taskId The ID of the task to update
 * @param taskData The updated task data
 * @returns Promise resolving to the updated task
 */
export const updateTask = async (userId: string, taskId: string, taskData: UpdateTaskRequest): Promise<Task> => {
  try {
    const response = await apiClient.put<Task>(`/api/${userId}/tasks/${taskId}`, taskData);
    return response.data;
  } catch (error: any) {
    console.error('Error updating task:', error);
    throw error;
  }
};

/**
 * Delete a task
 * @param userId The ID of the user who owns the task
 * @param taskId The ID of the task to delete
 * @returns Promise resolving when deletion is complete
 */
export const deleteTask = async (userId: string, taskId: string): Promise<void> => {
  try {
    await apiClient.delete(`/api/${userId}/tasks/${taskId}`);
  } catch (error: any) {
    console.error('Error deleting task:', error);
    throw error;
  }
};

/**
 * Toggle task completion status
 * @param userId The ID of the user who owns the task
 * @param taskId The ID of the task to update
 * @param completed The new completion status
 * @returns Promise resolving to the updated task
 */
export const toggleTaskCompletion = async (userId: string, taskId: string, completed: boolean): Promise<Task> => {
  try {
    const response = await apiClient.patch<Task>(`/api/${userId}/tasks/${taskId}/complete`, {
      completed
    });
    return response.data;
  } catch (error: any) {
    console.error('Error toggling task completion:', error);
    throw error;
  }
};

File: phase2/Todo app/backend/src/services/task_service.py
from sqlmodel import Session, select
from typing import Optional, List
from uuid import UUID
from ..models.task import Task, TaskBase
from ..models.user import User


class TaskService:
    def __init__(self, session: Session):
        self.session = session

    def create_task(self, user_id: UUID, task_data: TaskBase) -> Task:
        db_task = Task(
            title=task_data.title,
            description=task_data.description,
            completed=task_data.completed,
            user_id=user_id
        )
        self.session.add(db_task)
        self.session.commit()
        self.session.refresh(db_task)
        return db_task

    def get_tasks_by_user(self, user_id: UUID, completed: Optional[bool] = None) -> List[Task]:
        query = select(Task).where(Task.user_id == user_id)
        
        if completed is not None:
            query = query.where(Task.completed == completed)
            
        tasks = self.session.exec(query).all()
        return tasks

    def get_task_by_id(self, task_id: UUID, user_id: UUID) -> Optional[Task]:
        statement = select(Task).where(Task.id == task_id, Task.user_id == user_id)
        task = self.session.exec(statement).first()
        return task

    def update_task(self, task_id: UUID, user_id: UUID, task_data: TaskBase) -> Optional[Task]:
        task = self.get_task_by_id(task_id, user_id)
        if not task:
            return None
        
        # Update task attributes
        task.title = task_data.title
        task.description = task_data.description
        task.completed = task_data.completed
        
        self.session.add(task)
        self.session.commit()
        self.session.refresh(task)
        return task

    def delete_task(self, task_id: UUID, user_id: UUID) -> bool:
        task = self.get_task_by_id(task_id, user_id)
        if not task:
            return False
        
        self.session.delete(task)
        self.session.commit()
        return True

    def toggle_completion(self, task_id: UUID, user_id: UUID, completed: bool) -> Optional[Task]:
        task = self.get_task_by_id(task_id, user_id)
        if not task:
            return None
        
        task.completed = completed
        self.session.add(task)
        self.session.commit()
        self.session.refresh(task)
        return task